File: typescript/src/payment/api/paymentDetailsResponse.ts
// DTO di risposta per un Payment con il dettaglio delle sue transazioni.

import { PaymentDetails } from '../application/query/paymentDetails.js';
import { toTransactionResponse, type TransactionResponse } from './transactionResponse.js';

export type PaymentDetailsResponse = {
  readonly id: string;
  readonly clientReference: string;
  readonly amount: number;
  readonly status: string;
  readonly requestedAt: Date;
  readonly transactions: TransactionResponse[];
};

export function toPaymentDetailsResponse(details: PaymentDetails): PaymentDetailsResponse {
  return {
    id: details.id,
    clientReference: details.clientReference.toString(),
    amount: details.amount.value,
    status: details.status,
    requestedAt: details.requestedAt.value,
    transactions: details.transactions.map((t) => toTransactionResponse(t)),
  };
}

File: java/src/main/java/it/giannibombelli/wsc2026/payment/api/PaymentDetailsResponse.java
package it.giannibombelli.wsc2026.payment.api;

import it.giannibombelli.wsc2026.payment.application.query.PaymentDetails;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

public record PaymentDetailsResponse(
    UUID id,
    String clientReference,
    BigDecimal amount,
    String status,
    Instant requestedAt,
    List<TransactionResponse> transactions
) {
    public static PaymentDetailsResponse from(PaymentDetails details) {
        return new PaymentDetailsResponse(
            details.id(),
            details.clientReference().toString(),
            details.amount().value(),
            details.status().name(),
            details.requestedAt().value(),
            details.transactions().stream().map(PaymentDetailsResponse::from).toList()
        );
    }

    private static TransactionResponse from(PaymentDetails.TransactionDetail detail) {
        return new TransactionResponse(
            detail.id(),
            detail.provider().label(),
            detail.providerReference(),
            detail.amount().value(),
            detail.status().name(),
            detail.startedAt().value(),
            detail.completedAt() == null ? null : detail.completedAt().value()
        );
    }
}
